Add <mstd_xxx> C++ headers

kjbracey opened the pull request:
### Description

Restructure `mbed_cxxsupport.h` from #10868, to make it mirror the standard C library.

* `<mstd_xxxx>` is equivalent to C++14 or later `<xxxx>`
* Gets base toolchain header in to populate `namespace std` - some local implementation for ARM C 5.
* Imports all relevant stuff into `namespace mstd` - alternative implementations supplied where we want to bypass toolchain deficiencies. Eg `mstd::mutex` maps to `rtos::Mutex`.
* `namespace mstd` can also contain post-C++14 extensions not necessarily present in all toolchains.

Some C++11 library extensions and fixes, notably `invoke`, `<mstd_mutex>`.

`mbed::Atomic` in `platform/Atomic.h` moved to `mstd::atomic` in `<mstd_atomic>` to fit the general pattern.

Using separate namespace and filename gives us the ability to work around toolchain differences. In portable code, the following model could be used, assuming the choice between mbed OS and a fully C++11-compliant system.

    #ifdef TARGET_LIKE_MBED
    #include <mstd_atomic>
    #include <mstd_mutex>
    #else
    #include <atomic>
    #include <mutex>
    using namespace mstd = std;
    #endif

    using mstd::mutex;
     
    mutex m;
    mstd::atomic_int i;

This is a breaking change with respect to #10868 and #10274 now on master (reworking Atomic.h and mbed_cxxsupport.h), but they're new since 5.13. I'd like this change to go in before 5.14, so there's no visible released API breakage.

### Pull request type

    [ ] Fix
    [ ] Refactor
    [ ] Target update
    [X] Functionality change
    [ ] Docs update
    [ ] Test update
    [ ] Breaking change

### Reviewers

@pan-, @bulislaw 

### Release Notes

* A set of C++11/14 library support headers have been included in `platform/cxxsupport`. These include standard-C++-like APIs, eg `mstd::atomic` in `<mstd_atomic>`, to improve on implementations provided with toolchains. See `platform/cxxsupport/README.md` for more information.

kjbracey committed:
C++ invoke related stuff

For all compilers add post-C++14 stuff to namespace mbed:

* invoke, invoke_result, is_invocable, is_invocable_r,
  is_nothrow_invocable, is_nothrow_invocable_r, unwrap_reference,
  unwrap_ref_decay

For ARM C 5, add C++11 bits based on above to namespace std:

* result_of, reference_wrapper, ref, cref, mem_fn

kjbracey committed:
Split mbed_cxxsupport.h up - add namespace mstd

Regularise the C++ support glue, adding `<mstd_type_traits>` etc.

These include the base toolchain file, backfill `namespace std` as much
as possible for ARM C 5, and then arrange to create unified
`namespace mstd`, which can incorporate toolchain bypasses
(eg `mstd::swap` for ARM C 5, or `mstd::atomic`), and include local
implementations of post-ARM C++14 stuff.

All APIs in `namespace mstd` are intended to function as their
`namespace std` equivalent, and their presence there indicates they
are functional on all toolchains, and should be safe to use in
an Mbed OS build (including not unreasonable memory footprint).

See README.md for more info.

kjbracey committed:
Add TOOLCHAIN_ARMC5 label

We have some files that are needed for ARMC5 only.

kjbracey committed:
SingletonPtr: API extensions, make constexpr

* Adjust definition to make the default constructor `constexpr`.
  This permits use in classes that want lazy initialization and their
  own `constexpr` constructor, such as `mstd::mutex`.

* Add `get_no_init()` method to allow an explicit optimisation for
  paths that know they won be the first call (such as
  `mstd::mutex::unlock`).

* Add `destroy()` method to permit destruction of the contained object.
  (`SingletonPtr`'s destructor does not call its destructor - a cheat
  to omit destructors of static objects). Needed if using in a class
  that needs proper destruction.

kjbracey committed:
Add <mstd_mutex>

Add add-standard-as-possible version of C++11 <mutex>.

A lot of the stuff in there is generic, but the actual mutex classes and
call_once need to interface with the OS.

For those, they're not available in ARMC5 or IAR; retargetting would be
necessary for ARMC6 and GCC, and I've yet to investigate how whether
that's possible. So for now I'm using local implementations.

Although `Mutex` in principle could support `timed_mutex` and
`recursive_timed_mutex`, we don't have `chrono` for the time parameters,
so hold off for now.

For the generic stuff like mstd::unique_lock, they are aliased to
std::unique_lock where possible.

kjbracey committed:
astyle: exclude platform/cxxsupport

As something of a cheat, exclude platform/cxxsupport from astyle.

astyle really can't process the sort of template metaprogramming going
on in there; treat it as a "black box" like the toolchains' own
libraries.

kjbracey committed:
Move Atomic.h to <mstd_atomic>

`mbed::Atomic<T>` becomes `mstd::atomic<T>`, alongside the other
standard C++ library lookalikes.

kjbracey committed:
Make tools cope with no-extension headers

Treat files with no extension as headers, if they are in the cxxsupport
folder.

kjbracey committed:
Add ARMC5 unique_ptr

Standard library implementation should be fine for other toolchains.

kjbracey committed:
Add C++17 uninitialized storage ops

bulislaw commented on platform/SingletonPtr.h:
```diff
@@ -129,14 +134,48 @@ struct SingletonPtr {
+     *   A pointer to the singleton, or NULL if not
+     *   initialized.
+     */
+    T *get_no_init() const
```
Would it be worth to add an assert here?

kjbracey replied:
I think not here - I think returning NULL if it's not initialised is potentially useful API. You might want to perform some operation if it's active, but not start it.

Callers could maybe assert, but I don't think asserting ahead of a NULL dereference is that useful either - it will go bang with a pretty easy to debug backtrace.

bulislaw commented on tools/toolchains/arm.py:
```diff
@@ -160,9 +160,9 @@ def version_check(self):
     def _get_toolchain_labels(self):
         if getattr(self.target, "default_toolchain", "ARM") == "uARM":
-            return ["ARM", "ARM_MICRO"]
+            return ["ARM", "ARM_MICRO", "ARMC5"]
```
What does it do?

kjbracey replied:
Means we can optionally include stuff via a `TOOLCHAIN_ARMC5` folder. At the minute we only have the labels `TOOLCHAIN_ARM` and `TOOLCHAIN_ARMC6`. Used for the base `<array>`, `<initializer_list>` bits for ARM C 5.

(We only need to tiptoe around with namespace `mstd` and `<mstd_xxx>` for places where we are up against an existing header file; when we're supplying something ARM C 5 lacks altogether, we can just go straight in with the standard name in namespace `std`.)

bulislaw reviewed (approved):
Looks good, should be accompanied by good handbook docs to explain how and why to use it.

kjbracey committed:
Add mstd_xxx headers to unit tests

Stripped down headers that route through to `std` forms more - eg
no local implementation of `atomic` or `mutex`.

pan- commented on TESTS/mbed_platform/atomic/main.cpp:
```diff
@@ -28,6 +30,8 @@ using utest::v1::Case;
 
 namespace {
 
+using mstd::atomic;
```
`mbed.h` is included above, so there's `using namespace std` and `using mstd:atomic` that both brings `atomic` in the global namespace.

kjbracey replied:
This is actually pulling it into this unnamed namespace.

I previously claimed that a global-scope `using mstd::atomic` would override a more general `using namespace std` for `atomic`, and I'm sure I'd tested this, but seemed not to work here.

But putting it into an inner namespace like this does work - it searches innermost namespace first.

pan- replied:
I somehow missed the anonymous namespace declaration. Good to know.

pan- commented on UNITTESTS/target_h/platform/cxxsupport/mstd_algorithm:
```diff
@@ -0,0 +1,110 @@
+ *
+ * - provides <algorithm>
+ * - For ARM C 5, standard C++11/14 features:
+ *   - std::min, std::max for std::initializer_list
```
Why not have <mstd_initializer_list> for initialiser list ?

kjbracey replied:
Good general question - for consistency, it would indeed make sense to have `mstd_xxx` for all headers we think are good, to "approve" them, and to allow a single `using namespace mstd`, even if we currently don't need any adaptation, and it would just copy everything over. If you want to fill them in, I wouldn't object, at least for "not-too-bloaty-for-embedded" stuff.

pan- commented on UNITTESTS/target_h/platform/cxxsupport/mstd_algorithm:
```diff
@@ -0,0 +1,110 @@
+ *   - std::find_if_not
+ *   - std::equal (2-range forms)
+ *   - std::copy_n, std::move, std::move_backward
+ *   - mstd::min, mstd::max constexpr replacements
```
How do we get constexpr replacement if `std::min` and `std::max` are directly imported ?

kjbracey replied:
As `mstd::min` and `mstd::max`.  For ARM C 5 those are these replacement implementations, rather than being aliased to `std::min` like for the other toolchains.

pan- commented on UNITTESTS/target_h/platform/cxxsupport/mstd_cstddef:
```diff
@@ -0,0 +1,54 @@
+#define MSTD_CONSTEXPR_OBJ_14 constexpr
+#else
+#define MSTD_CONSTEXPR_FN_14 inline
+#define MSTD_CONSTEXPR_OBJ_14 const
```
I'm not sure to agree with that definition; in C++14 a constexpr object can invoke non const member function in a constexpr context. In C++11; constexpr equals const.

kjbracey replied:
`constexpr` implies const method in a C++11, yes, but I don't think that's an issue here. That would be used as:

    MSTD_CONSTEXPR_FN_14 int my_complicated_method() /* const or not */ {
          // some-multi-line thing
         return whatever;
    }

Flipping from constexpr in C++14 to inline in C++11 seems right to me - either member or non-member.

Or were you misunderstanding `OBJ` - that was intended for a constexpr object:

    static MSTD_CONSTEXPR_OBJ_14 int npos = -1;

pan- replied:
I was thinking of the following case: 

```c++
struct Foo { 
    constexpr Foo() { }

    // not a const member function!
    MSTD_CONSTEXPR_FN_14 int get_something() { /* ... */ }
};

static MSTD_CONSTEXPR_OBJ_14 Foo foo;

void foo() { 
   // error call of a non const function on a const object
   auto v = foo.get_something();
}
```

kjbracey replied:
Seems fine to me? In C++14, the function is `constexpr`, not const, and it fails to compile.

In C++11, the function is `inline`, not const, and it fails to compile.

pan- commented on platform/cxxsupport/README.md:
```diff
@@ -0,0 +1,65 @@
+User code should normally be able to include C++14 headers and get
+most expected functionality. (But bear in mind that many C++ library
+features like streams and STL containers are quite heavy and may
+not be appropriate for small embnedded use).
```
`embnedded ` => `embedded `

pan- commented on platform/cxxsupport/README.md:
```diff
@@ -0,0 +1,65 @@
+## C++ support ##
```
We need a page like this in the handbook.

kjbracey replied:
Yes, already noted, and I'll be working on it when I get back from holiday. It needs a more complete easy-to-read table listing what's available and what the tweaks are. Probably for both `mstd` and `std` equivalents alongside each other.

pan- commented on platform/cxxsupport/TOOLCHAIN_ARMC5/array:
```diff
@@ -0,0 +1,171 @@
+        return const_reverse_iterator(begin());
+    }
+};
+
```
lexicographic comparison operators are missing.

pan- commented on platform/cxxsupport/mstd_algorithm:
```diff
@@ -0,0 +1,385 @@
+template <typename T>
+T max(initializer_list<T> il)
+{
+    return *std::max_element(begin(il), end(il));
```
How efficient that is compared to a recursive approach ?

kjbracey replied:
Last thing we want to do in embedded space is be recursive. And surely we can assume `max_element` is already efficient?

Recursion is certainly a thing for "make it constexpr for C++11", but if it ever did fall back to run-time, we'd be doomed. So generally I want to steer clear of constexpr if it means potential run-time recursion. We don't get to have 2 implementations :(

On the whole, I'm not /that/ interested in tuning ARM C 5 for performance, it's more just trying to get it off the ground and not break the build.

But there are places where constexpr is useful - we're currently missing `aligned_union`, which might be wanted as a basis for `variant`, and its implementation might want to do a constexpr `max` on a pack of `sizeof` via an initializer list. So might end up changing that one. I guess we can assume no-one is passing around non-constant initializer-lists, right? (They're not supposed to be able to construct them themselves).

pan- replied:
Thanks for the explanations. I should have clarified my question. I wanted to know if the compiler is able to do that at compile time when the list of Ts is known. It was curiosity more than anything else 😄 .

kjbracey replied:
Ah, right. I believe it won't do it at compile time, no. Because the existing `max_element` is a straightforward loop, and that's beyond the compile-time capabilities of the compiler.

But a recursive constexpr implementation would be possible, but would be less efficient if it was ever computed for non-constant arguments.

C++11's constexpr restrictions forces you to write weird algorithms to achieve constexprness. You don't generally want to do it, but for the specific case of initializer_lists which are normally (but not necessarily) constant, you might want to do it.

pan- commented on platform/cxxsupport/mstd_cstddef:
```diff
@@ -0,0 +1,75 @@
+#define MSTD_CONSTEXPR_OBJ_14 constexpr
+#else
+#define MSTD_CONSTEXPR_FN_14 inline
+#define MSTD_CONSTEXPR_OBJ_14 const
```
See comment above.

pan- commented on platform/cxxsupport/mstd_functional:
```diff
@@ -0,0 +1,508 @@
+// [refwrap]
+template <typename T>
+class reference_wrapper {
+    T &FUN(T &x) noexcept { return x; }
```
Question, why use `FUN` as identifier ? `__fun` or `_Fun` where not good enough ?

kjbracey replied:
It came from the revised-in-2017 standard text (https://cplusplus.github.io/LWG/issue2993). We're not actually using it anyway, as that version seems to be too advanced for ARM C 5. I'll remove the dead code.

pan- commented on platform/cxxsupport/mstd_functional:
```diff
@@ -0,0 +1,508 @@
+}
+
+template <typename F>
+impl::not_fn_t<F> not_fn_t(F&& f)
```
Do you mean `not_fn` as the function name ?

kjbracey replied:
Spot the person rushing to get this in before holidays. (To ensure we don't have a breaking change versus a shipped mbed_cxxsupport.h)

To some extent this is justified for the bulk of the code being for ARMC5 only, and it's not-officially-supported status, but yes, testing clearly needed.

pan- commented on platform/cxxsupport/mstd_memory:
```diff
@@ -0,0 +1,616 @@
+template<
+    class T,
+    class D = default_delete<T>
+> class unique_ptr : public impl::deleter_store<D>
```
Glad to see that class getting in.

kjbracey replied:
Yes, but sadly in use, it increases code size, at least in my tests.

You'd think it was "free", but the compiler doesn't seem to be as good at optimising out shuffling. For example, given:

     if (!ptr) {
         ptr = make_unique<Thing>(whatever);
     }

I've seen ARMC5 at least put in a pointless delete-of-current-pointer-contents on the assignment (but we just tested that the pointer is currently empty), and then a pointless delete of the temporary returned by make_unique (we know we just moved out of it).

Avoiding make_unique helped a bit, but still not as good as doing it by hand.

Maybe other compilers (and their std::unique_ptr) do better?

kjbracey replied:
Much as I want to believe the C++ guys' claims that C++ is as efficient as C, I keep seeing cases like this where it's just a bit too hard for the compiler to totally eliminate the generic stuff, and it does get bigger. At this embedded scale of tens of kilobytes of ROM and RAM, the impact is real...

Lots of little micro things you wouldn't think of.

Plus the annoyances like what I hit with `SingletonPtr` here - took me a number of hours to understand the initialization rules enough to comprehend the "mixed zero and constant initialization" problem I was hitting and deal with it to get them back out of the final link.

Was watching this while debugging that one: https://www.youtube.com/watch?v=D7Sd8A6_fYU

Made me a little grumpy - yes, there are misconceptions about C++ for embedded use, but it does also have real issues. And yes, sure, stronger type checking can move more debugging to compilation time, but often that debugging is much harder...

pan- replied:
It would be sweet if our documentations were explaining which C++ features is safe to use and what are the impacts on code size/speed (I'd love to write that). I hope most of the code won't turn into crazy template so debug ability should be okay; too bad that compilers doesn't have attributes to treat constructors as _literals_. It would be really useful for basic building blocks such as `Callback`. 

> Much as I want to believe the C++ guys' claims that C++ is as efficient as C, I keep seeing cases like this where it's just a bit too hard for the compiler to totally eliminate the generic stuff, and it does get bigger. At this embedded scale of tens of kilobytes of ROM and RAM, the impact is real...

I'd be interested to measure the impact on a real world application. Not that I doubt there is one (there will be one as long as there's no destructive moves) but an accurate picture would help us create guidelines to our community of users and internal developers.

pan- commented on platform/cxxsupport/mstd_mutex:
```diff
@@ -0,0 +1,425 @@
+};
+
+// [thread.mutex.recursive]
+class recursive_mutex : public _Mutex_base {
```
any plan to add the `timed_` versions ?

kjbracey replied:
We need `chrono` for ARMC5 first. That feels like a separate project.

If licensing permitted just lifting the ARMC6 implementation...?

kjbracey replied:
This is one case where ARMC5 is a significant blocker. It would be easy to bring chrono in to use for the other 3 toolchains, but any attempt to keep ARMC5 building blocks it. Could make all relevant APIs conditional on `ifndef __CC_ARM`, but don't fancy that either.

pan- reviewed (commented):
Awesome work; I skimmed through files (it still took me few hours) and this looks solid. Some tests would help for future maintenance.
